<font size="7"><p style="text-align: center;">**Gérer ses données avec python**</font></p>
<br><br><br>
<img src="img\pandas.svg">

## <u>1. Importer des données avec Pandas</u>
Pandas offre une panoplie de connecteurs pour importer des données (CSV, Excel, SQL, hdfs, etc.).

data source: https://www.kaggle.com/rashikrahmanpritom/heart-attack-analysis-prediction-dataset

In [95]:
import pandas as pd 

In [96]:
!pip install pyodbc

In [97]:
!pip install lxml

In [98]:
!pip install pandas

In [99]:
!pip install sqlalchemy

In [100]:
import pyodbc
import numpy as np
import pandas as pd
from copy import copy
from sqlalchemy import create_engine

In [101]:
data = pd.read_csv('heart.csv')

# La méthode head() permet de n'afficher que les premières lignes (5 par défaut)
# Ici on en affiche 10
data.head()
data.tail()

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0

Il existe 2 structures de donnée dans pandas:
- les **Series**: correspond à une colonne, à un numpy array avec un type de donnée unique
- les **DataFrame**: correspond à une collection de **Series**

## <u>2. Comment manipuler ses données avec pandas ?</u>

**2.1. Indexation et colonnes**

In [102]:
# Avec le slicing, comme pour les listes python
# En ne sélectionnant qu'une seule ligne
data_sliced = data.loc[74]
display(data_sliced)
# Selectionner plusieurs lignes avec une list ou une array
data_sliced2 = data.loc[[10,20,30]]
display(data_sliced2)
# Selectionner sur base de l'index et des noms de colonnes
data_sliced3 = data.loc[[10,20,30],['age','chol']]
display(data_sliced3)
data_sliced4 = data[['age','chol']].loc[1:5]
display(data_sliced4)

display(data_sliced3.dtypes)
data_sliced4.dtypes

age          43.0
sex           0.0
cp            2.0
trtbps      122.0
chol        213.0
fbs           0.0
restecg       1.0
thalachh    165.0
exng          0.0
oldpeak       0.2
slp           1.0
caa           0.0
thall         2.0
output        1.0
Name: 74, dtype: float64

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
10   54    1   0     140   239    0        1       160     0      1.2    2   
20   59    1   0     135   234    0        1       161     0      0.5    1   
30   41    0   1     105   198    0        1       168     0      0.0    2   

    caa  thall  output  
10    0      2       1  
20    0      3       1  
30    1      2       1

age  chol
10   54   239
20   59   234
30   41   198

age  chol
1   37   250
2   41   204
3   56   236
4   57   354
5   57   192

age     int64
chol    int64
dtype: object

age     int64
chol    int64
dtype: object

In [103]:
# En sélectionnant un range
data_sliced = data.loc[74:80]
data_sliced

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
74   43    0   2     122   213    0        1       165     0      0.2    1   
75   55    0   1     135   250    0        0       161     0      1.4    1   
76   51    1   2     125   245    1        0       166     0      2.4    1   
77   59    1   1     140   221    0        1       164     1      0.0    2   
78   52    1   1     128   205    1        1       184     0      0.0    2   
79   58    1   2     105   240    0        0       154     1      0.6    1   
80   41    1   2     112   250    0        1       179     0      0.0    2   

    caa  thall  output  
74    0      2       1  
75    0      2       1  
76    0      2       1  
77    0      2       1  
78    0      2       1  
79    0      3       1  
80    0      2       1

In [104]:
data.iloc[0:5,0:3]

age  sex  cp
0   63    1   3
1   37    1   2
2   41    0   1
3   56    1   1
4   57    0   0

In [106]:
# Il est également possible de ne sélectionner qu'une colonne en particulier
# Avec les crochets et la colonne voulue
data_sliced = data[['age','chol']]
data_sliced
data_chol = data['chol']
print (data_sliced)
data_chol

     age  chol
0     63   233
1     37   250
2     41   204
3     56   236
4     57   354
..   ...   ...
298   57   241
299   45   264
300   68   193
301   57   131
302   57   236

[303 rows x 2 columns]


0      233
1      250
2      204
3      236
4      354
      ... 
298    241
299    264
300    193
301    131
302    236
Name: chol, Length: 303, dtype: int64

In [107]:
# Avec plusieurs colonnes
# A noter les doubles crochets
data_sliced = data[['age','output']].loc[54:60]
data_sliced

age  output
54   63       1
55   52       1
56   48       1
57   45       1
58   34       1
59   57       1
60   71       1

**2.2. Boolean indexing**

Nous avons vu jusqu'ici comment sélectionner des parties de notre DataFrame avec les indexes ou le nom des colonnes.

Nous pourrions également faire une sélection basée sur des conditions. 

C'est ce que l'on appelle le **boolean indexing**.

In [108]:
# On ne sélectionne que les personnes plus agées que 58 ans
data_sliced = data[data['age'] > 58]
data_sliced
data_sliced2= data[data['age']>58].loc[0:50]
print(data_sliced2.loc[38])
print(data_sliced2.iloc[10])

age          65.0
sex           0.0
cp            2.0
trtbps      155.0
chol        269.0
fbs           0.0
restecg       1.0
thalachh    148.0
exng          0.0
oldpeak       0.8
slp           2.0
caa           0.0
thall         2.0
output        1.0
Name: 38, dtype: float64
age          65.0
sex           0.0
cp            2.0
trtbps      155.0
chol        269.0
fbs           0.0
restecg       1.0
thalachh    148.0
exng          0.0
oldpeak       0.8
slp           2.0
caa           0.0
thall         2.0
output        1.0
Name: 38, dtype: float64


On peut également mettre plusieurs conditions.

**Attention:** les opérateurs logiques "and" et "or" classiques de Python ne fonctionnent pas.

Il faut impérativement utiliser les opérateurs **&** et **|** et mettre des **()** autour des conditions.

In [109]:
data_sliced = data[(data['age'] > 58) & (data['output'] == 1)]
data_sliced
data_sliced2 = data[(data['age'] > 58) & (data['output'] == 1)].loc[1:100]
data_sliced2

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
13   64    1   3     110   211    0        0       144     1      1.8    1   
17   66    0   3     150   226    0        1       114     0      2.6    0   
19   69    0   3     140   239    0        1       151     0      1.8    2   
20   59    1   0     135   234    0        1       161     0      0.5    1   
23   61    1   2     150   243    1        1       137     1      1.0    1   
25   71    0   1     160   302    0        1       162     0      0.4    2   
26   59    1   2     150   212    1        1       157     0      1.6    2   
28   65    0   2     140   417    1        0       157     0      0.8    2   
31   65    1   0     120   177    0        1       140     0      0.4    2   
38   65    0   2     155   269    0        1       148     0      0.8    2   
39   65    0   2     160   360    0        0       151     0      0.8    2   
51   66    1   0     120   302    0        0       151     0      0.4    1   
52   62    1   2     130   231    0        1       146     0      1.8    1   
54   63    0   2     135   252    0        0       172     0      0.0    2   
60   71    0   2     110   265    1        0       130     0      0.0    2   
69   62    0   0     124   209    0        1       163     0      0.0    2   
77   59    1   1     140   221    0        1       164     1      0.0    2   
82   60    0   2     102   318    0        1       160     0      0.0    2   
85   67    0   2     115   564    0        0       160     0      1.6    1   
86   68    1   2     118   277    0        1       151     0      1.0    2   
96   62    0   0     140   394    0        0       157     0      1.2    1   

    caa  thall  output  
13    0      2       1  
17    0      2       1  
19    2      2       1  
20    0      3       1  
23    0      2       1  
25    2      2       1  
26    0      2       1  
28    1      2       1  
31    0      3       1  
38    0      2       1  
39    0      2       1  
51    0      2       1  
52    3      3       1  
54    0      2       1  
60    1      2       1  
69    0      2       1  
77    0      2       1  
82    1      2       1  
85    0      3       1  
86    1      3       1  
96    0      2       1

Il est possible de créer de nouvelles colonnes de manière à enrichir vos données.

In [110]:
# Ici on souhaite créer une colonne basée sur la colonne age de manière à obtenir des catégories
# On l'instancie d'abord à 'unknown'
data['age_cat'] = 'unknown'
data.head()
del data['age_cat']

In [111]:
# Nous allons utiliser le boolean indexing pour modifier la catégorie
# Attention à utiliser .loc lorsque l'on chercher un modifier des valeurs
data.loc[data['age'] <= 18, 'age_cat'] = 'child'
data.loc[(data['age'] > 18) & (data['age'] < 66), 'age_cat'] = 'adult'
data.loc[data['age'] > 65, 'age_cat'] = 'senior'
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1   adult  
1      0      2       1   adult  
2      0      2       1   adult  
3      0      2       1   adult  
4      0      2       1   adult  
..   ...    ...     ...     ...  
298    0      3       0   adult  
299    0      3       0   adult  
300    2      3       0  senior  
301    1      3       0   adult  
302    1      2       0   adult  

[303 rows x 15 columns]

### Supprimer une colonne

In [112]:
del data['age_cat']

In [113]:
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0     63    1   3     145   233    1        0       150     0      2.3    0   
1     37    1   2     130   250    0        1       187     0      3.5    0   
2     41    0   1     130   204    0        0       172     0      1.4    2   
3     56    1   1     120   236    0        1       178     0      0.8    2   
4     57    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   57    0   0     140   241    0        1       123     1      0.2    1   
299   45    1   3     110   264    0        1       132     0      1.2    1   
300   68    1   0     144   193    1        1       141     0      3.4    1   
301   57    1   0     130   131    0        1       115     1      1.2    1   
302   57    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
0      0      1       1  
1      0      2       1  
2      0      2       1  
3      0      2       1  
4      0      2       1  
..   ...    ...     ...  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0  

[303 rows x 14 columns]

## <u>3. Opérations mathématiques</u>
- Pandas supporte bien évidemment toutes les opérations mathématiques de base
- Pandas supporte également toutes les opérations de numpy

In [114]:
data_math = copy(data)
# Addition d'une constante à l'entièreté d'une colonne
data_math['age'] += 100
# identique a data_math['age'] = data_math['age'] +100
data_math

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    163    1   3     145   233    1        0       150     0      2.3    0   
1    137    1   2     130   250    0        1       187     0      3.5    0   
2    141    0   1     130   204    0        0       172     0      1.4    2   
3    156    1   1     120   236    0        1       178     0      0.8    2   
4    157    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298  157    0   0     140   241    0        1       123     1      0.2    1   
299  145    1   3     110   264    0        1       132     0      1.2    1   
300  168    1   0     144   193    1        1       141     0      3.4    1   
301  157    1   0     130   131    0        1       115     1      1.2    1   
302  157    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
0      0      1       1  
1      0      2       1  
2      0      2       1  
3      0      2       1  
4      0      2       1  
..   ...    ...     ...  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0  

[303 rows x 14 columns]

In [ ]:
# On peut également additionner les colonnes entre elles
data_math['age'] = data_math['age']+data_math['chol']
data_math

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    396    1   3     145   233    1        0       150     0      2.3    0   
1    387    1   2     130   250    0        1       187     0      3.5    0   
2    345    0   1     130   204    0        0       172     0      1.4    2   
3    392    1   1     120   236    0        1       178     0      0.8    2   
4    511    0   0     120   354    0        1       163     1      0.6    2   
..   ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298  398    0   0     140   241    0        1       123     1      0.2    1   
299  409    1   3     110   264    0        1       132     0      1.2    1   
300  361    1   0     144   193    1        1       141     0      3.4    1   
301  288    1   0     130   131    0        1       115     1      1.2    1   
302  393    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output  
0      0      1       1  
1      0      2       1  
2      0      2       1  
3      0      2       1  
4      0      2       1  
..   ...    ...     ...  
298    0      3       0  
299    0      3       0  
300    2      3       0  
301    1      3       0  
302    1      2       0  

[303 rows x 14 columns]

On peut retrouver toutes les fonctions mathématiques de numpy ici:

https://numpy.org/doc/stable/reference/routines.math.html

In [ ]:
data_math['age'] = np.log(data_math['age'])
data_math
data_test = np.sum(data_math['age'])
data_test

np.float64(1526.387480233499)

## <u>4. Agrégations et statistiques</u>
Tout comme avec numpy, il est possible d'agréger nos données selon un axe spécifique.

Pandas fournit également quelque statistiques de base.

In [ ]:
# Par défaut axis=0
data.sum()

age         16473.0
sex           207.0
cp            293.0
trtbps      39882.0
chol        74618.0
fbs            45.0
restecg       160.0
thalachh    45343.0
exng           99.0
oldpeak       315.0
slp           424.0
caa           221.0
thall         701.0
output        165.0
dtype: float64

In [ ]:
data.sum(axis=1)

0      600.3
1      614.5
2      554.4
3      598.8
4      701.6
       ...  
298    567.2
299    561.2
300    558.4
301    442.2
302    602.0
Length: 303, dtype: float64

In [ ]:
# Imaginons que nous devions calculer la moyenne pour chaque colonne
print(data.shape)
print(data.sum()/data.shape[0])
print(data.sum(axis=1)/data.shape[1])

(303, 14)
age          54.366337
sex           0.683168
cp            0.966997
trtbps      131.623762
chol        246.264026
fbs           0.148515
restecg       0.528053
thalachh    149.646865
exng          0.326733
oldpeak       1.039604
slp           1.399340
caa           0.729373
thall         2.313531
output        0.544554
dtype: float64
0      42.878571
1      43.892857
2      39.600000
3      42.771429
4      50.114286
         ...    
298    40.514286
299    40.085714
300    39.885714
301    31.585714
302    43.000000
Length: 303, dtype: float64


In [ ]:
# Plus simple comme ceci:
print(data[['age','chol']].mean())
data[['age','chol']].loc[1:20].mean()

age      54.366337
chol    246.264026
dtype: float64


age      53.75
chol    246.95
dtype: float64

In [ ]:
# Lien de corrélation
data.corr()

age       sex        cp    trtbps      chol       fbs  \
age       1.000000 -0.098447 -0.068653  0.279351  0.213678  0.121308   
sex      -0.098447  1.000000 -0.049353 -0.056769 -0.197912  0.045032   
cp       -0.068653 -0.049353  1.000000  0.047608 -0.076904  0.094444   
trtbps    0.279351 -0.056769  0.047608  1.000000  0.123174  0.177531   
chol      0.213678 -0.197912 -0.076904  0.123174  1.000000  0.013294   
fbs       0.121308  0.045032  0.094444  0.177531  0.013294  1.000000   
restecg  -0.116211 -0.058196  0.044421 -0.114103 -0.151040 -0.084189   
thalachh -0.398522 -0.044020  0.295762 -0.046698 -0.009940 -0.008567   
exng      0.096801  0.141664 -0.394280  0.067616  0.067023  0.025665   
oldpeak   0.210013  0.096093 -0.149230  0.193216  0.053952  0.005747   
slp      -0.168814 -0.030711  0.119717 -0.121475 -0.004038 -0.059894   
caa       0.276326  0.118261 -0.181053  0.101389  0.070511  0.137979   
thall     0.068001  0.210041 -0.161736  0.062210  0.098803 -0.032019   
output   -0.225439 -0.280937  0.433798 -0.144931 -0.085239 -0.028046   

           restecg  thalachh      exng   oldpeak       slp       caa  \
age      -0.116211 -0.398522  0.096801  0.210013 -0.168814  0.276326   
sex      -0.058196 -0.044020  0.141664  0.096093 -0.030711  0.118261   
cp        0.044421  0.295762 -0.394280 -0.149230  0.119717 -0.181053   
trtbps   -0.114103 -0.046698  0.067616  0.193216 -0.121475  0.101389   
chol     -0.151040 -0.009940  0.067023  0.053952 -0.004038  0.070511   
fbs      -0.084189 -0.008567  0.025665  0.005747 -0.059894  0.137979   
restecg   1.000000  0.044123 -0.070733 -0.058770  0.093045 -0.072042   
thalachh  0.044123  1.000000 -0.378812 -0.344187  0.386784 -0.213177   
exng     -0.070733 -0.378812  1.000000  0.288223 -0.257748  0.115739   
oldpeak  -0.058770 -0.344187  0.288223  1.000000 -0.577537  0.222682   
slp       0.093045  0.386784 -0.257748 -0.577537  1.000000 -0.080155   
caa      -0.072042 -0.213177  0.115739  0.222682 -0.080155  1.000000   
thall    -0.011981 -0.096439  0.206754  0.210244 -0.104764  0.151832   
output    0.137230  0.421741 -0.436757 -0.430696  0.345877 -0.391724   

             thall    output  
age       0.068001 -0.225439  
sex       0.210041 -0.280937  
cp       -0.161736  0.433798  
trtbps    0.062210 -0.144931  
chol      0.098803 -0.085239  
fbs      -0.032019 -0.028046  
restecg  -0.011981  0.137230  
thalachh -0.096439  0.421741  
exng      0.206754 -0.436757  
oldpeak   0.210244 -0.430696  
slp      -0.104764  0.345877  
caa       0.151832 -0.391724  
thall     1.000000 -0.344029  
output   -0.344029  1.000000

Nous avons vu comment créer des catégories. Comment calculer la moyenne par exemple en fonction de ces catégories?

Avec la fonction **groupby** comme en SQL.


In [115]:
# Il suffit de spécifier la colonne
data.loc[data['age'] <= 18, 'age_cat'] = 'child'
data.loc[(data['age'] > 18) & (data['age'] < 66), 'age_cat'] = 'adult'
data.loc[data['age'] > 65, 'age_cat'] = 'senior'
data.groupby(['age_cat','output']).mean(numeric_only=True)

age       sex        cp      trtbps        chol  \
age_cat output                                                          
adult   0       55.073770  0.811475  0.467213  133.573770  251.286885   
        1       50.608108  0.594595  1.358108  128.716216  239.743243   
senior  0       68.250000  0.937500  0.562500  140.687500  249.562500   
        1       68.941176  0.294118  1.529412  134.411765  263.882353   

                     fbs   restecg    thalachh      exng   oldpeak       slp  \
age_cat output                                                                 
adult   0       0.147541  0.450820  140.065574  0.540984  1.604918  1.180328   
        1       0.141892  0.601351  160.628378  0.148649  0.549324  1.608108   
senior  0       0.250000  0.437500  131.750000  0.625000  1.437500  1.062500   
        1       0.117647  0.529412  139.647059  0.058824  0.876471  1.470588   

                     caa     thall  
age_cat output                      
adult   0       1.098361  2.549180  
        1       0.324324  2.128378  
senior  0       1.687500  2.500000  
        1       0.705882  2.058824

In [ ]:
data.groupby(['age_cat','output']).agg({'age':['mean','max','min'],
                                        'chol':'max'})

age                            chol     
             mean            max     min      max     
output          0          1   0   1   0   1    0    1
age_cat                                               
adult    55.07377  50.608108  65  65  35  29  409  417
senior   68.25000  68.941176  77  76  66  66  322  564

In [ ]:
# Nous n'avons pas la catégorie child, pourquoi?
# Value counts permet d'inventorier nos valeurs catégorielles
data['age_cat'].value_counts()

age_cat
adult     270
senior     33
Name: count, dtype: int64

In [ ]:
# Vérifions
print(data['age'].min())
data['age'].max()


29


np.int64(77)

In [ ]:
# La fonction describe permet d'obtenir une vision très macro de vos données
data.describe(include='all')

age         sex          cp      trtbps        chol  \
count   303.000000  303.000000  303.000000  303.000000  303.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean     54.366337    0.683168    0.966997  131.623762  246.264026   
std       9.082101    0.466011    1.032052   17.538143   51.830751   
min      29.000000    0.000000    0.000000   94.000000  126.000000   
25%      47.500000    0.000000    0.000000  120.000000  211.000000   
50%      55.000000    1.000000    1.000000  130.000000  240.000000   
75%      61.000000    1.000000    2.000000  140.000000  274.500000   
max      77.000000    1.000000    3.000000  200.000000  564.000000   

               fbs     restecg    thalachh        exng     oldpeak  \
count   303.000000  303.000000  303.000000  303.000000  303.000000   
unique         NaN         NaN         NaN         NaN         NaN   
top            NaN         NaN         NaN         NaN         NaN   
freq           NaN         NaN         NaN         NaN         NaN   
mean      0.148515    0.528053  149.646865    0.326733    1.039604   
std       0.356198    0.525860   22.905161    0.469794    1.161075   
min       0.000000    0.000000   71.000000    0.000000    0.000000   
25%       0.000000    0.000000  133.500000    0.000000    0.000000   
50%       0.000000    1.000000  153.000000    0.000000    0.800000   
75%       0.000000    1.000000  166.000000    1.000000    1.600000   
max       1.000000    2.000000  202.000000    1.000000    6.200000   

               slp         caa       thall      output age_cat  
count   303.000000  303.000000  303.000000  303.000000     303  
unique         NaN         NaN         NaN         NaN       2  
top            NaN         NaN         NaN         NaN   adult  
freq           NaN         NaN         NaN         NaN     270  
mean      1.399340    0.729373    2.313531    0.544554     NaN  
std       0.616226    1.022606    0.612277    0.498835     NaN  
min       0.000000    0.000000    0.000000    0.000000     NaN  
25%       1.000000    0.000000    2.000000    0.000000     NaN  
50%       1.000000    0.000000    2.000000    1.000000     NaN  
75%       2.000000    1.000000    3.000000    1.000000     NaN  
max       2.000000    4.000000    3.000000    1.000000     NaN

In [ ]:
print(data.shape)
print(data.size)
print(data.dtypes)
print(data.columns)



In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 15 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trtbps    303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalachh  303 non-null    int64  
 8   exng      303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slp       303 non-null    int64  
 11  caa       303 non-null    int64  
 12  thall     303 non-null    int64  
 13  output    303 non-null    int64  
 14  age_cat   303 non-null    object 
dtypes: float64(1), int64(13), object(1)
memory usage: 35.6+ KB


## <u>5. Comment gérer les valeurs manquantes:</u>

In [ ]:
# D'abord modifions notre dataset de sorte à avoir des données manquantes
# nan correspond à Null, aux valeurs manquantes
# Ici l'âge pour toutes les personnes de moins de 58 ans sera à NaN
data.loc[data['age'] < 58, 'age'] = np.nan
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    63.0    1   3     145   233    1        0       150     0      2.3    0   
1     NaN    1   2     130   250    0        1       187     0      3.5    0   
2     NaN    0   1     130   204    0        0       172     0      1.4    2   
3     NaN    1   1     120   236    0        1       178     0      0.8    2   
4     NaN    0   0     120   354    0        1       163     1      0.6    2   
..    ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298   NaN    0   0     140   241    0        1       123     1      0.2    1   
299   NaN    1   3     110   264    0        1       132     0      1.2    1   
300  68.0    1   0     144   193    1        1       141     0      3.4    1   
301   NaN    1   0     130   131    0        1       115     1      1.2    1   
302   NaN    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1   adult  
1      0      2       1   adult  
2      0      2       1   adult  
3      0      2       1   adult  
4      0      2       1   adult  
..   ...    ...     ...     ...  
298    0      3       0   adult  
299    0      3       0   adult  
300    2      3       0  senior  
301    1      3       0   adult  
302    1      2       0   adult  

[303 rows x 15 columns]

In [ ]:
# D'abord il faut les identifier
# isna() permet de renvoyer True ou False si les données sont manquantes ou non
# Ensuite sum() additionne le tout. False étant == 0 et True == 1 on peut en déduire le nombre d'élément manquant
data.isna().sum()

age         180
sex           0
cp            0
trtbps        0
chol          0
fbs           0
restecg       0
thalachh      0
exng          0
oldpeak       0
slp           0
caa           0
thall         0
output        0
age_cat       0
dtype: int64

In [ ]:
# fillna() permet de remplacer des valeurs manquantes par une valeur choisie
# ici on choisir de remplacer par la moyenne
data['age'] = data['age'].fillna(data['age'].mean())
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
0    63.0    1   3     145   233    1        0       150     0      2.3    0   
1    63.0    1   2     130   250    0        1       187     0      3.5    0   
2    63.0    0   1     130   204    0        0       172     0      1.4    2   
3    63.0    1   1     120   236    0        1       178     0      0.8    2   
4    63.0    0   0     120   354    0        1       163     1      0.6    2   
..    ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
298  63.0    0   0     140   241    0        1       123     1      0.2    1   
299  63.0    1   3     110   264    0        1       132     0      1.2    1   
300  68.0    1   0     144   193    1        1       141     0      3.4    1   
301  63.0    1   0     130   131    0        1       115     1      1.2    1   
302  63.0    0   1     130   236    0        0       174     0      0.0    1   

     caa  thall  output age_cat  
0      0      1       1   adult  
1      0      2       1   adult  
2      0      2       1   adult  
3      0      2       1   adult  
4      0      2       1   adult  
..   ...    ...     ...     ...  
298    0      3       0   adult  
299    0      3       0   adult  
300    2      3       0  senior  
301    1      3       0   adult  
302    1      2       0   adult  

[303 rows x 15 columns]

On pourrait également supprimer les valeurs manquantes.

Attention à ne pas trop supprimer de lignes ou de colonnes.

In [ ]:
data.loc[data['age'] == 63, 'age'] = np.nan
data = data.dropna()
data

age  sex  cp  trtbps  chol  fbs  restecg  thalachh  exng  oldpeak  slp  \
13   64.0    1   3     110   211    0        0       144     1      1.8    1   
14   58.0    0   3     150   283    1        0       162     0      1.0    2   
16   58.0    0   2     120   340    0        1       172     0      0.0    2   
17   66.0    0   3     150   226    0        1       114     0      2.6    0   
19   69.0    0   3     140   239    0        1       151     0      1.8    2   
..    ...  ...  ..     ...   ...  ...      ...       ...   ...      ...  ...   
291  58.0    1   0     114   318    0        2       140     0      4.4    0   
292  58.0    0   0     170   225    1        0       146     1      2.8    1   
293  67.0    1   2     152   212    0        0       150     0      0.8    1   
297  59.0    1   0     164   176    1        0        90     0      1.0    1   
300  68.0    1   0     144   193    1        1       141     0      3.4    1   

     caa  thall  output age_cat  
13     0      2       1   adult  
14     0      2       1   adult  
16     0      2       1   adult  
17     0      2       1  senior  
19     2      2       1  senior  
..   ...    ...     ...     ...  
291    3      1       0   adult  
292    2      1       0   adult  
293    0      3       0  senior  
297    2      1       0   adult  
300    2      3       0  senior  

[114 rows x 15 columns]

Nous sommes passés de 303 lignes à 114, ce qui fait beaucoup.

Il serait plus sage de supprimer la colonne age plutôt que de supprimer chaque ligne où il y a une valeur manquante.

## <u>6. Comment join plusieurs datasets </U>

Fonctionnement identique au join SQL

Illustation avec le pd.read_clipboard()

PIB: https://fr.wikipedia.org/wiki/Produit_int%C3%A9rieur_brut

Population: https://fr.wikipedia.org/wiki/Liste_des_pays_par_population

In [ ]:
pib = pd.read_clipboard()
pib

Pays  2014  2015  2016  2017  \
0                    Drapeau du Luxembourg Luxembourg   283   282   278   269   
1                   Drapeau des États-Unis États-Unis   148   148   142   140   
2                     Drapeau de la Belgique Belgique   121   121   120   118   
3                    Drapeau de l'Allemagne Allemagne   127   124   125   124   
4                         Drapeau de la France France   108   107   106   104   
5   Drapeau de l’Union européenne Union européenne...   100   100   100   100   
6                  Drapeau du Royaume-Uni Royaume-Uni   111   111   109   108   
7                          Drapeau de l'Italie Italie    98    97    99    98   
8                        Drapeau de l'Espagne Espagne    90    91    92    93   
9                              Drapeau du Japon Japon   107   107   100    97   
10                      Drapeau de la Pologne Pologne    67    69    69    69   
11                      Drapeau de la Turquie Turquie    65    68    66    66   

    2018  2019  2020  2021  2022  2023  
0    260   252   259   266   256   239  
1    140   134   135   135   134   136  
2    118   118   118   120   120   118  
3    124   121   123   120   117   115  
4    104   106   105   103   100   101  
5    100   100   100   100   100   100  
6    105   104   102   102    99    98  
7     97    97    94    96    98    97  
8     91    91    83    84    85    88  
9     95    89    90    86    82    84  
10    71    73    76    77    80    80  
11    63    59    61    61    67    73

In [ ]:

pib['Pays'] = pib['Pays'].apply(lambda x: x.split()[-1])
pib

Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023
0    Luxembourg   283   282   278   269   260   252   259   266   256   239
1    États-Unis   148   148   142   140   140   134   135   135   134   136
2      Belgique   121   121   120   118   118   118   118   120   120   118
3     Allemagne   127   124   125   124   124   121   123   120   117   115
4        France   108   107   106   104   104   106   105   103   100   101
5         pays)   100   100   100   100   100   100   100   100   100   100
6   Royaume-Uni   111   111   109   108   105   104   102   102    99    98
7        Italie    98    97    99    98    97    97    94    96    98    97
8       Espagne    90    91    92    93    91    91    83    84    85    88
9         Japon   107   107   100    97    95    89    90    86    82    84
10      Pologne    67    69    69    69    71    73    76    77    80    80
11      Turquie    65    68    66    66    63    59    61    61    67    73

In [ ]:
pib = pib.drop(5)


In [ ]:
pib

Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023
0    Luxembourg   283   282   278   269   260   252   259   266   256   239
1    États-Unis   148   148   142   140   140   134   135   135   134   136
2      Belgique   121   121   120   118   118   118   118   120   120   118
3     Allemagne   127   124   125   124   124   121   123   120   117   115
4        France   108   107   106   104   104   106   105   103   100   101
6   Royaume-Uni   111   111   109   108   105   104   102   102    99    98
7        Italie    98    97    99    98    97    97    94    96    98    97
8       Espagne    90    91    92    93    91    91    83    84    85    88
9         Japon   107   107   100    97    95    89    90    86    82    84
10      Pologne    67    69    69    69    71    73    76    77    80    80
11      Turquie    65    68    66    66    63    59    61    61    67    73

In [ ]:
population = pd.read_html('https://fr.wikipedia.org/wiki/Liste_des_pays_par_population')
population = population[0]
population

Rang (2021) Pays ou territoire Population au 1er juillet 2021  \
0             -              Monde                  7 909 295 151   
1             1           Chine[a]                  1 425 893 465   
2             2               Inde                  1 407 563 842   
3             3      États-Unis[b]                    336 997 624   
4             4          Indonésie                    273 753 191   
..          ...                ...                            ...   
233           -         Montserrat                          4 417   
234           -          Malouines                          3 764   
235         199               Niue                          1 937   
236           -            Tokelau                          1 849   
237         200            Vatican                            511   

    Population projetée au 1er janvier 2025  
0                             8 155 601 099  
1                             1 424 864 149  
2                             1 448 210 906  
3                               342 715 948  
4                               280 914 242  
..                                      ...  
233                                   4 362  
234                                   3 812  
235                                   1 935  
236                                   1 925  
237                                     529  

[238 rows x 4 columns]

In [ ]:
population['Pays ou territoire'] = population['Pays ou territoire'].str.replace(r'\[.*\]', '', regex=True).str.strip()

In [ ]:
population

Rang (2021) Pays ou territoire Population au 1er juillet 2021  \
0             -              Monde                  7 909 295 151   
1             1              Chine                  1 425 893 465   
2             2               Inde                  1 407 563 842   
3             3         États-Unis                    336 997 624   
4             4          Indonésie                    273 753 191   
..          ...                ...                            ...   
233           -         Montserrat                          4 417   
234           -          Malouines                          3 764   
235         199               Niue                          1 937   
236           -            Tokelau                          1 849   
237         200            Vatican                            511   

    Population projetée au 1er janvier 2025  
0                             8 155 601 099  
1                             1 424 864 149  
2                             1 448 210 906  
3                               342 715 948  
4                               280 914 242  
..                                      ...  
233                                   4 362  
234                                   3 812  
235                                   1 935  
236                                   1 925  
237                                     529  

[238 rows x 4 columns]

In [ ]:
print(list(population))
print(list(pib))
population = pib.merge(population, how='left', left_on='Pays', right_on='Pays ou territoire')
population

['Rang (2021)', 'Pays ou territoire', 'Population au 1er juillet 2021', 'Population projetée au 1er janvier 2025']
['Pays', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022', '2023']


Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023  \
0    Luxembourg   283   282   278   269   260   252   259   266   256   239   
1    États-Unis   148   148   142   140   140   134   135   135   134   136   
2      Belgique   121   121   120   118   118   118   118   120   120   118   
3     Allemagne   127   124   125   124   124   121   123   120   117   115   
4        France   108   107   106   104   104   106   105   103   100   101   
5   Royaume-Uni   111   111   109   108   105   104   102   102    99    98   
6        Italie    98    97    99    98    97    97    94    96    98    97   
7       Espagne    90    91    92    93    91    91    83    84    85    88   
8         Japon   107   107   100    97    95    89    90    86    82    84   
9       Pologne    67    69    69    69    71    73    76    77    80    80   
10      Turquie    65    68    66    66    63    59    61    61    67    73   

   Rang (2021) Pays ou territoire Population au 1er juillet 2021  \
0          165         Luxembourg                        639 321   
1            3         États-Unis                    336 997 624   
2           81           Belgique                     11 611 419   
3           19          Allemagne                     83 408 554   
4           22             France                     64 531 444   
5           21        Royaume-Uni                     67 281 039   
6           25             Italie                     59 240 329   
7           30            Espagne                     47 486 935   
8           11              Japon                    124 612 530   
9           38            Pologne                     38 307 726   
10          18            Turquie                     84 775 404   

   Population projetée au 1er janvier 2025  
0                                  664 929  
1                              342 715 948  
2                               11 730 450  
3                               83 228 580  
4                               64 943 206  
5                               68 072 841  
6                               58 609 511  
7                               47 448 540  
8                              122 298 912  
9                               39 868 923  
10                              86 479 284

In [ ]:
population = population.rename(columns={'Rang (2021)':'Rang population'})
population

Pays  2014  2015  2016  2017  2018  2019  2020  2021  2022  2023  \
0    Luxembourg   283   282   278   269   260   252   259   266   256   239   
1    États-Unis   148   148   142   140   140   134   135   135   134   136   
2      Belgique   121   121   120   118   118   118   118   120   120   118   
3     Allemagne   127   124   125   124   124   121   123   120   117   115   
4        France   108   107   106   104   104   106   105   103   100   101   
5   Royaume-Uni   111   111   109   108   105   104   102   102    99    98   
6        Italie    98    97    99    98    97    97    94    96    98    97   
7       Espagne    90    91    92    93    91    91    83    84    85    88   
8         Japon   107   107   100    97    95    89    90    86    82    84   
9       Pologne    67    69    69    69    71    73    76    77    80    80   
10      Turquie    65    68    66    66    63    59    61    61    67    73   

   Rang population Pays ou territoire Population au 1er juillet 2021  \
0              165         Luxembourg                        639 321   
1                3         États-Unis                    336 997 624   
2               81           Belgique                     11 611 419   
3               19          Allemagne                     83 408 554   
4               22             France                     64 531 444   
5               21        Royaume-Uni                     67 281 039   
6               25             Italie                     59 240 329   
7               30            Espagne                     47 486 935   
8               11              Japon                    124 612 530   
9               38            Pologne                     38 307 726   
10              18            Turquie                     84 775 404   

   Population projetée au 1er janvier 2025  
0                                  664 929  
1                              342 715 948  
2                               11 730 450  
3                               83 228 580  
4                               64 943 206  
5                               68 072 841  
6                               58 609 511  
7                               47 448 540  
8                              122 298 912  
9                               39 868 923  
10                              86 479 284

## <u>7. connection db </U>

<br>

Créer une connecteur sur base d'une URL avec comme information:
- le SGBD
- username+password
- adresse réseau
- le nom de la db à laquelle se connecter

E.g.:
- postgresql://username:password@localhost:5432/db_name # localhost=127.0.0.1
- mssql+pyodbc://username:password@localhost:1433/db_name
- mssql+pyodbc://server_name/db_name?trusted_connection=yes # auth windows
- mssql+pyodbc://server_name/db_name?trusted_connection=yes&driver=ODBC+Driver+17+for+SQL+Server # Avec driver ODBC spécifique

In [ ]:
sgbd = 'mssql+pyodbc'
server_name = r'GEORGES\DATA'
db_name = 'DBSlide'
driver = 'ODBC+Driver+17+for+SQL+Server'

url = f'{sgbd}://{server_name}/{db_name}?trusted_connection=yes&driver={driver}'
con = create_engine(url)

sql_query = "SELECT * FROM student WHERE year_result > 15"
data = pd.read_sql(sql_query,con)

con.dispose() # release con

data

student_id first_name last_name birth_date     login  section_id  \
0           5      Kevin     Bacon 1958-07-08    kbacon        1120   
1           6        Kim  Basinger 1953-12-08  kbasinge        1310   
2           8      Julia   Roberts 1967-10-28  jroberts        1120   
3          11       Andy    Garcia 1956-04-12   agarcia        1110   
4          18   Jennifer    Garner 1972-04-17   jgarner        1120   
5          25      Halle     Berry 1966-08-14    hberry        1320   

   year_result course_id  
0           16         0  
1           19         0  
2           17         0  
3           19         0  
4           18         0  
5           18    EG2210